Week 7/Solution.java

```diff
@@ -0,0 +1,175 @@
+			//checks if product is available
+			String check = checkProduct(code);
+			//if available
+			if(!(check.equals("Sold out")||check.equals("Wrong code"))){
+				//variable to store the change that must be generated
+				int aux = money - i.getProducts().get(code).getPrice();
+				if(aux<0) return false;
+				//if it was able to generate the change
+				if(m.generateChange(aux)){
+					//dispense the product
+					i.sellProduct(code);
+					//Stores the money
+					m.addDenomination(money,1);
+					return true;
+				}
+				//cancels the sale
+				return false;
+			}else{
+				return false;
+			}
+		}
+	}
+	class Money{
+		//amount of coins per denomination
+		private HashMap<Integer,Integer> current;
+		//available denominations
+		private ArrayList<Integer> coins;
+		//constructors
+		Money(){
+			current = new HashMap<Integer,Integer>();
+		}
+		Money(HashMap<Integer,Integer> current,ArrayList<Integer> coins){
+			this.current = current;
+			this.coins = coins;
+		}
+		//adds new denomination
+		void addDenomination(int value,int amount){
+			if(current.containsKey(value)){
+				current.put(value,current.get(value)+amount);
+			}else{
+				current.put(value,amount);
+				coins.add(value);
+			}
+			Collections.sort(coins);
+		}
+		HashMap<Integer,Integer> getCurrent(){
+			return current;
+		}
+		boolean generateChange(int money){
+			//recursive method to check if it is possible to generate the change
+			for(int i = coins.size()-1;i>=0; i--){
+				int cur_coin = coins.get(i);
+				//if the money needed is equals to the current coin it returns true and deletes the coins
+				if(cur_coin==money){
+					if(current.get(cur_coin)>1){
+						current.put(cur_coin,current.get(cur_coin)-1);
+					}else{
+						current.remove(cur_coin);
+						coins.remove(i);
+					}
+					return true;
+				}
+				if(cur_coin<money){
+					//if this path is valid, it deletes the coins used
+					if(generateChange(money-coins.get(i))){
+						if(current.get(cur_coin)>1){
+							current.put(cur_coin,current.get(cur_coin)-1);
+						}else{
+							current.remove(cur_coin);
+							coins.remove(i);
+						}
+						return true;
+					}
+				}
+			}
+			return false;
+		}
+	}
+	class Inventory{
+		private HashMap<String,Product> products;
+		Inventory(){
+			products =  new HashMap<String,Product>();
+		}
+		Inventory(HashMap<String,Product> products){
+			this.products = products;
+		}
+		//Stores new product
+		void addProduct(String code, int price,int amount){
+			if(products.containsKey(code)){
+				products.get(code).setAmount(products.get(code).getAmount()+amount);
+			}else{
+				Product p = new Product(price,amount);
+				products.put(code,p);
+			}
+		}
+		//takes the sold item from the inventory
+		void sellProduct(String code){
+			products.get(code).setAmount(products.get(code).getAmount()-1);
+		}
+		HashMap<String,Product> getProducts(){
+			return products;
+		}
+	}
+	class Product{
+		private int price;
+		private int amount;
```

Comment:
Consider storing the same product in multiple slots. A more flexible design might be to allow inventory to have a mapping of Product to its amount.